### PR TITLE
Address issue #184, observations not being updated when changed

### PR DESCRIPTION
**Pull request by dustinburson**
* Address issue #184, observations not being updated when changed.
 - Defect was due to using two different references to the same object for the change comparison.
 - Updated CodeValueFhirTemplateProcessor.MergeObservation to perform a DeepCopy of the passed in Observation parameter. The copy is what is returned now by the function to ensure modifications are made to a new instance.
 - Moved the logic to set the status of the Observation from the MergeObservation function to the R4FhirImportService.SaveObservationAsync logic.  Without the move, the first time a created observation is "updated" will always result in the FHIR resource being saved even if nothing else changed because we were always setting the status to amended.
 - Updated unit tests to account for new logic.
 - Added extension methods for some common scenarios
* Rename FhirClientValidator to FhirClientValidatorExtensions to match other extension class names
* Rename ServiceCollectionExtrensions to FhirClientExtensions to match purpose.
* Merge content of HttpClientBuilderRegistraionExtensions into FhirClientExtensions.
* Remove registering ITelemetryLogger singleton with FhirClient. ITelemetryLogger is resolved from the service provider now.  This was preventing local debugging with the Azure Function.
* Update AddAuthenticationHandler to use the registered singletons for the auth service instead of creating new instances.

### CODE REVIEW

File: src/lib/Microsoft.Health.Fhir.R4.Ingest/Service/R4FhirImportService.cs [hunk 1]
```diff
@@ -113,8 +113,8 @@ public virtual async Task<string> SaveObservationAsync(ILookupTemplate<IFhirTemp
                     // Merge the new data with the existing Observation.
                     var mergedObservation = MergeObservation(config, existingObservation, observationGroup);
 
-                    // Check to see if there are any changes after merging.
-                    if (mergedObservation.IsExactly(existingObservation))
+                    // Check to see if there are any changes after merging and update the Status to ammended if changed.
```
Comment (rogordon01) [inline src/lib/Microsoft.Health.Fhir.R4.Ingest/Service/R4FhirImportService.cs:116]: Nit. Typo in `amend`

File: src/lib/Microsoft.Health.Extensions.Fhir.R4/ResourceExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+{
+    public static class ResourceExtensions
+    {
+        /// <summary>
+        /// Performs full deep copy of the resource.
+        /// </summary>
+        /// <typeparam name="TResource">Type of resource to return.</typeparam>
+        /// <param name="resource">Resource to copy.</param>
+        /// <returns>New resource object with the contents of the original.</returns>
+        public static TResource FullCopy<TResource>(this TResource resource)
```
Comment (rogordon01) [inline src/lib/Microsoft.Health.Extensions.Fhir.R4/ResourceExtensions.cs:19]: I only see this referenced in unit tests. Is that intended?

File: src/lib/Microsoft.Health.Fhir.R4.Ingest.Templates/Template/CodeValueFhirTemplateProcessor.cs [hunk 3]
```diff
@@ -128,42 +128,42 @@ protected override Observation MergeObservationImpl(CodeValueFhirTemplate templa
             EnsureArg.IsNotNull(grp, nameof(grp));
             EnsureArg.IsNotNull(existingObservation, nameof(existingObservation));
 
-            existingObservation.Status = ObservationStatus.Amended;
+            var mergedObservation = existingObservation.DeepCopy() as Observation;
```
Comment (rogordon01) [inline src/lib/Microsoft.Health.Fhir.R4.Ingest.Templates/Template/CodeValueFhirTemplateProcessor.cs:131]: Should this be using the new extension method?:

```.net
var mergedObservation = existingObservation.FullCopy()
```

**Review by rogordon01 — APPROVED** [review]

**Comment by rogordon01** [comment]
Have you deployed and tested this change out in a personal cluster?

File: src/lib/Microsoft.Health.Fhir.R4.Ingest.Templates/Template/CodeValueFhirTemplateProcessor.cs (hunk 3 above)
Comment (dustinburson) [reply src/lib/Microsoft.Health.Fhir.R4.Ingest.Templates/Template/CodeValueFhirTemplateProcessor.cs:131]: Unfortunately we can't right now.  This project is .NET 2.0 standard where as the Extenstions,Fhir.R4 project is .NET 6.0 and can't be downgraded due to dependencies.  This project needs to stay .NET Standard till we can move our RP to .NET 6.0.

I could have introduced a completely separate project just for this method but it seemed like overkill.

File: src/lib/Microsoft.Health.Extensions.Fhir.R4/ResourceExtensions.cs (hunk 2 above)
Comment (dustinburson) [reply src/lib/Microsoft.Health.Extensions.Fhir.R4/ResourceExtensions.cs:19]: Covered in other comment but yes for now it is due to project compatibility issues.

File: src/lib/Microsoft.Health.Fhir.R4.Ingest/Service/R4FhirImportService.cs (hunk 1 above)
Comment (dustinburson) [reply src/lib/Microsoft.Health.Fhir.R4.Ingest/Service/R4FhirImportService.cs]: Thanks, fixed in latest version.

**Comment by dustinburson** [comment]
> Have you deployed and tested this change out in a personal cluster?

Not yet in a personal cluster.  I ran the Azure Function locally pointed to a remote FHIR Service using these instructions https://github.com/microsoft/iomt-fhir/blob/main/docs/Debugging.md#example-post-body  testing various update and create scenarios.

**Comment by rogordon01** [comment]
Running it as an Azure Funtions works too 😊

From: Dustin Burson ***@***.***>
Date: Friday, April 15, 2022 at 1:07 PM
To: microsoft/iomt-fhir ***@***.***>
Cc: Rob Gordon ***@***.***>, Review requested ***@***.***>
Subject: Re: [microsoft/iomt-fhir] Address issue #184, observations not being updated when changed (PR #185)

Have you deployed and tested this change out in a personal cluster?

Not yet in a personal cluster. I ran the Azure Function locally pointed to a remote FHIR Service using these instructions https://github.com/microsoft/iomt-fhir/blob/main/docs/Debugging.md#example-post-body<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fmicrosoft%2Fiomt-fhir%2Fblob%2Fmain%2Fdocs%2FDebugging.md%23example-post-body&data=05%7C01%7CRob.Gordon%40microsoft.com%7Ca8724709f1a6456eea0108da1f1ba038%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637856500758965649%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=%2BH4XdiswauxWzi3qyTGXTknwQL10DeG2JnH3xe3nzKc%3D&reserved=0> testing various update and create scenarios.

**Pull request merged**